### PR TITLE
fix sdk not exist error

### PR DESCRIPTION
**Pull request by Varorbc**

### CODE REVIEW

**Comment by ErikEJ** [comment]
What problem are you trying to solve here?

what does: `dotnet --version` return on your machine?

**Comment by Varorbc** [comment]
My machine has installed 3.1 and 6.0 SDKs, and the command line returns 6.0

**Comment by ErikEJ** [comment]
Why not just change to:

```
            return result.StartsWith("3.", StringComparison.OrdinalIgnoreCase)
                     || result.StartsWith("5.", StringComparison.OrdinalIgnoreCase)
                     || result.StartsWith("6.", StringComparison.OrdinalIgnoreCase)
```

**Comment by Varorbc** [comment]
Because 6.0 is in preview state, and if the user only installs 6.0, it cannot be used normally

**Comment by ErikEJ** [comment]
Should it not be --list-runtimes ? the full SDK is not required here.

Also, what does " it cannot be used normally" mean? What errors do you get?

**Comment by ErikEJ** [comment]
Sorry, --list-runtimes wont work

**Comment by ErikEJ** [comment]
where did you download the .NET 6 SDK from?

**Comment by Varorbc** [comment]
> where did you download the .NET 6 SDK from?

https://aka.ms/dotnet/net6/dev/Sdk/dotnet-sdk-win-x64.exe

**Comment by ErikEJ** [comment]
>> Also, what does " it cannot be used normally" mean? What errors do you get?

**Comment by Varorbc** [comment]
> > > Also, what does " it cannot be used normally" mean? What errors do you get?

can ef core 5.0 run normally on 6.0? It's just a guess that I haven't tested yet

**Comment by ErikEJ** [comment]
>>can ef core 5.0 run normally on 6.0?

The question is, can the efreveng tool run on .NET6 - and I dont see any reason why not.

Why not test my proposal?   https://github.com/ErikEJ/EFCorePowerTools/pull/696#issuecomment-753895047

**Comment by Varorbc** [comment]
> > > can ef core 5.0 run normally on 6.0?
> 
> The question is, can the efreveng tool run on .NET6 - and I dont see any reason why not.
> 
> Why not test my proposal? [#696 (comment)](https://github.com/ErikEJ/EFCorePowerTools/pull/696#issuecomment-753895047)

OK, I'll do a test tomorrow

**Comment by Varorbc** [comment]
I got a "table list error: unable to launch external process:" error

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs [hunk 1]
```diff
@@ -123,13 +125,26 @@ private bool IsDotnetInstalled()
...
 
             var result = RunProcess(startInfo);
 
-            return result.StartsWith("3.", StringComparison.OrdinalIgnoreCase)
-                || result.StartsWith("5.", StringComparison.OrdinalIgnoreCase);
+            var isInstalled = false;
+            var sdks = result.Split(new string[] { "\r\n" }, StringSplitOptions.RemoveEmptyEntries);
+            foreach (var item in sdks)
+            {
+                isInstalled = useEFCore5 ?
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: The tool just needs either dotnet 3 or dotnet 5, useEFCore5 is not relevant here.

**Review by ErikEJ — CHANGES_REQUESTED** [review]

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs (hunk 1 above)
Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: If the user uses EFCore3.0, the 3.0 SDK needs to be installed on the machine. If the user uses EFCore5.0, the 5.0 SDK needs to be installed. It will not work without the corresponding SDK

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: EF Core 5 works fine with .NET 3.1! and this is for running the reverse engineer tool, not for runtime concerns

Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: https://github.com/ErikEJ/EFCorePowerTools/blob/master/src/GUI/efpt30.core/efpt30.core.csproj#L5

```
.NET SDK (反映任何 global.json):
 Version:   5.0.101
 Commit:    d05174dc5a

运行时环境:
 OS Name:     Windows
 OS Version:  10.0.19042
 OS Platform: Windows
 RID:         win10-x86
 Base Path:   C:\Program Files (x86)\dotnet\sdk\5.0.101\

Host (useful for support):
  Version: 6.0.0-alpha.1.20560.10
  Commit:  72b7d236ad

.NET SDKs installed:
  5.0.101 [C:\Program Files (x86)\dotnet\sdk]
  6.0.100-alpha.1.20562.2 [C:\Program Files (x86)\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 5.0.1 [C:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0-alpha.1.20526.6 [C:\Program Files (x86)\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 5.0.1 [C:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0-alpha.1.20560.10 [C:\Program Files (x86)\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 5.0.1 [C:\Program Files (x86)\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0-alpha.1.20560.7 [C:\Program Files (x86)\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
```

image

image

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: https://github.com/ErikEJ/EFCorePowerTools/blob/master/src/GUI/efpt30.core/efpt30.core.csproj#L5 has nothing to do with reverse engineering

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: efreveng and efrevneg50 are the reverse engineering tools

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: Just remove the useEFCore5 check

Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: If useEFCore5 is checked, whether to use efreveng50 reverse engineering, if so, there is no 5.0 runtime on the machine (currently it is unlikely, because VS already includes 5.0 SDK), how does efreveng50 work normally?

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: efreveng50 is a .NET Core 3.1 app (as you can see in the .csproj)

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs [hunk 2]
```diff
@@ -79,7 +81,7 @@ private List<TableModel> GetTablesInternal(string arguments)
         {
             if (!IsDotnetInstalled())
             {
-                throw new Exception($"Reverse engineer error: Unable to launch 'dotnet' version 3 or newer. Do you have it installed?");
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:82]: Revert this change

**Review by ErikEJ — CHANGES_REQUESTED** [review]

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs (hunk 1 above)
Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: Sorry, I missed it. Is it enough to check whether 3.1 is installed?

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: 3, 5 or 6 will work (as in the previous implementation) - newer runtimes can run older binaries.

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: You can also change the check to "3.1" if you want, yes

Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: As shown in the figure above, there is no SDK for 3.1 on the machine, and the tool cannot work normally

**Review by ErikEJ — CHANGES_REQUESTED** [review]

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs (hunk 1 above)
Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: Are you testing with your modified code?

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: Looking at the error message, that does not look like it is the case.

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs [hunk 3]
```diff
@@ -123,13 +125,23 @@ private bool IsDotnetInstalled()
...
 
             var result = RunProcess(startInfo);
 
-            return result.StartsWith("3.", StringComparison.OrdinalIgnoreCase)
-                || result.StartsWith("5.", StringComparison.OrdinalIgnoreCase);
+            var isInstalled = false;
+            var sdks = result.Split(new string[] { "\r\n" }, StringSplitOptions.RemoveEmptyEntries);
+            foreach (var item in sdks)
+            {
+                isInstalled = item.StartsWith("3.1", StringComparison.OrdinalIgnoreCase);
```
Comment (ErikEJ) [inline src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: add `|| item.StartsWith("5."`

**Review by ErikEJ — CHANGES_REQUESTED** [review]

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs (hunk 3 above)
Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: https://github.com/ErikEJ/EFCorePowerTools/pull/696#discussion_r551754781

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: Basically implement my initial comment: https://github.com/ErikEJ/EFCorePowerTools/pull/696#issuecomment-753895047

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: Yes, but was that a test of your branch ?? - the error message shown indicates that it was not.

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs (hunk 1 above)
Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs]: Not my modified version, but the version downloaded from the market for testing

**Comment by Varorbc** [comment]
Is it possible to upgrade the target frameworks of efreveng and efreveng50 to net5.0, because currently VS contains 5.0sdk

**Comment by ErikEJ** [comment]
>> Is it possible to upgrade the target frameworks of efreveng and efreveng50 to net5.0, because currently VS contains 5.0sdk

VS also conatins the 3.1 SDK, so I see no reason for that breaking change, also the tool supports both EF Core 3.1 and EF Core 5.0

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs (hunk 1 above)
Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs]: Well, add the check for 5.0 and show me the result of testing your local changes then.

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs (hunk 3 above)
Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: Not my modified version, but the version downloaded from the market for testing

Comment (ErikEJ) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: Sorry to repeat my self: add `|| item.StartsWith("5."` and test your modififcations

**Comment by Varorbc** [comment]
> VS also conatins the 3.1 SDK, so I see no reason for that breaking change, also the tool supports both EF Core 3.1 and EF Core 5.0

image
16.8 only contains 5.0 sdk

**Comment by ErikEJ** [comment]
>> 16.8 only contains 5.0 sdk

No, that just means that this is the first version containing this SDK. 
See https://dotnet.microsoft.com/download/dotnet-core/3.1

**Comment by Varorbc** [comment]
> > > 16.8 only contains 5.0 sdk
> 
> No, that just means that this is the first version containing this SDK.
> See https://dotnet.microsoft.com/download/dotnet-core/3.1

3.1 SDK from VS no exists

image

File: src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs (hunk 3 above)
Comment (Varorbc) [reply src/GUI/EFCorePowerTools/Handlers/ReverseEngineer/EfRevEngLauncher.cs:137]: image

**Comment by ErikEJ** [comment]
image

Just just need to add it.

**Comment by Varorbc** [comment]
> image
> 
> You just need to add it.

This is an option, not a required one. I think it's better to use the default SDK of vs as the target framework.

**Comment by Varorbc** [comment]
This is an option, not a required one. I think it's better to use the default SDK of vs as the target framework.



> image
> 
> You just need to add it.

This is an option, not a required one. I think it's better to use the default SDK of vs as the target framework.

**Comment by ErikEJ** [comment]
Not an option, as long as I support EF Core 3.1 - your fix is wrong btw, I am working on a correct fix.
